File: java-app/src/main/java/com/k8smanager/service/NatsService.java
package com.k8smanager.service;

import io.nats.client.Connection;
import io.nats.client.Nats;
import io.nats.client.Message;
import io.nats.client.Subscription;
import io.nats.client.Dispatcher;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;

import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.concurrent.CompletableFuture;
import java.util.Map;
import java.util.HashMap;

public class NatsService {
    private Connection natsConnection;
    private ObjectMapper objectMapper;
    private KubernetesService kubernetesService;
    private Dispatcher dispatcher;
    
    public NatsService(KubernetesService kubernetesService) throws Exception {
        this.kubernetesService = kubernetesService;
        this.objectMapper = new ObjectMapper();
        this.objectMapper.registerModule(new JavaTimeModule());
        
        // Connect to NATS server
        String natsUrl = System.getenv("NATS_URL");
        if (natsUrl == null) {
            natsUrl = "nats://nats-service:4222";
        }
        
        this.natsConnection = Nats.connect(natsUrl);
        System.out.println("Connected to NATS at: " + natsUrl);
        
        // Create dispatcher for handling messages
        this.dispatcher = natsConnection.createDispatcher();
        
        // Set up subscriptions
        setupSubscriptions();
        
        // Start publishing cluster info periodically
        startPeriodicPublishing();
    }
    
    private void setupSubscriptions() throws Exception {
        // Subscribe to requests for cluster info
        dispatcher.subscribe("k8s.cluster.info.request", (msg) -> {
            try {
                System.out.println("Received cluster info request");
                Map<String, Object> clusterInfo = kubernetesService.getClusterInfo();
                String response = objectMapper.writeValueAsString(clusterInfo);
                
                if (msg.getReplyTo() != null) {
                    natsConnection.publish(msg.getReplyTo(), response.getBytes(StandardCharsets.UTF_8));
                }
            } catch (Exception e) {
                System.err.println("Error handling cluster info request: " + e.getMessage());
            }
        });
        
        // Subscribe to general events
        dispatcher.subscribe("k8s.events", (msg) -> {
            String data = new String(msg.getData(), StandardCharsets.UTF_8);
            System.out.println("Received k8s event: " + data);
        });
        
        // Subscribe to commands
        dispatcher.subscribe("k8s.commands", (msg) -> {
            String command = new String(msg.getData(), StandardCharsets.UTF_8);
            System.out.println("Received command: " + command);
            handleCommand(command, msg.getReplyTo());
        });
        
        System.out.println("NATS subscriptions established");
    }
    
    private void handleCommand(String command, String replyTo) {
        try {
            Map<String, Object> response = new HashMap<>();
            
            switch (command) {
                case "status":
                    response.put("status", "healthy");
                    response.put("service", "k8s-manager");
                    break;
                case "cluster-info":
                    response = kubernetesService.getClusterInfo();
                    break;
                default:
                    response.put("error", "Unknown command: " + command);
            }
            
            if (replyTo != null) {
                String responseJson = objectMapper.writeValueAsString(response);
                natsConnection.publish(replyTo, responseJson.getBytes(StandardCharsets.UTF_8));
            }
        } catch (Exception e) {
            System.err.println("Error handling command: " + e.getMessage());
        }
    }
    
    private void startPeriodicPublishing() {
        CompletableFuture.runAsync(() -> {
            while (!Thread.currentThread().isInterrupted()) {
                try {
                    Thread.sleep(30000); // 30 seconds
                    
                    Map<String, Object> status = new HashMap<>();
                    status.put("timestamp", System.currentTimeMillis());
                    status.put("service", "k8s-manager");
                    status.put("status", "running");
                    
                    String statusJson = objectMapper.writeValueAsString(status);
                    natsConnection.publish("app.status", statusJson.getBytes(StandardCharsets.UTF_8));
                    
                    // Also publish cluster metrics
                    Map<String, Object> clusterInfo = kubernetesService.getClusterInfo();
                    Map<String, Object> metrics = new HashMap<>();
                    metrics.put("podCount", clusterInfo.get("podCount"));
                    metrics.put("deploymentCount", clusterInfo.get("deploymentCount"));
                    metrics.put("timestamp", System.currentTimeMillis());
                    
                    String metricsJson = objectMapper.writeValueAsString(metrics);
                    natsConnection.publish("k8s.metrics", metricsJson.getBytes(StandardCharsets.UTF_8));
                    
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    break;
                } catch (Exception e) {
                    System.err.println("Error publishing periodic updates: " + e.getMessage());
                }
            }
        });
    }
    
    public void publishEvent(String subject, Object data) throws Exception {
        String json = objectMapper.writeValueAsString(data);
        natsConnection.publish(subject, json.getBytes(StandardCharsets.UTF_8));
    }
    
    public void close() throws Exception {
        if (natsConnection != null) {
            natsConnection.close();
        }
    }
}

File: python-k8s-manager/services/nats_service.py
import asyncio
import nats
import json
import threading
import time
from typing import Optional, Dict, Any
from concurrent.futures import ThreadPoolExecutor

class NatsService:
    """Service for NATS messaging"""
    
    def __init__(self, nats_url: str = "nats://nats-service:4222"):
        self.nats_url = nats_url
        self.nc: Optional[nats.NATS] = None
        self.loop: Optional[asyncio.AbstractEventLoop] = None
        self.executor = ThreadPoolExecutor(max_workers=2)
        self._running = False
    
    def start(self) -> bool:
        """Start NATS service in background thread"""
        try:
            # Start event loop in separate thread
            self._running = True
            self.executor.submit(self._run_async_loop)
            
            # Wait a moment for connection
            time.sleep(2)
            return self.nc is not None
            
        except Exception as e:
            print(f"Failed to start NATS service: {e}")
            return False
    
    def _run_async_loop(self):
        """Run asyncio event loop in thread"""
        self.loop = asyncio.new_event_loop()
        asyncio.set_event_loop(self.loop)
        
        try:
            self.loop.run_until_complete(self._connect_and_run())
        except Exception as e:
            print(f"NATS async loop error: {e}")
        finally:
            self.loop.close()
    
    async def _connect_and_run(self):
        """Connect to NATS and keep connection alive"""
        try:
            self.nc = await nats.connect(servers=[self.nats_url])
            print(f"Connected to NATS at {self.nats_url}")
            
            # Set up subscriptions
            await self._setup_subscriptions()
            
            # Keep connection alive
            while self._running:
                await asyncio.sleep(1)
                
        except Exception as e:
            print(f"NATS connection error: {e}")
        finally:
            if self.nc:
                await self.nc.close()
    
    async def _setup_subscriptions(self):
        """Set up NATS subscriptions"""
        try:
            # Subscribe to commands
            await self.nc.subscribe("k8s.commands", cb=self._handle_command)
            
            # Subscribe to events
            await self.nc.subscribe("k8s.events", cb=self._handle_event)
            
            print("NATS subscriptions established")
            
        except Exception as e:
            print(f"Error setting up NATS subscriptions: {e}")
    
    async def _handle_command(self, msg):
        """Handle command messages"""
        try:
            command = msg.data.decode()
            print(f"Received command: {command}")
            
            response = {"source": "python-k8s-manager", "command": command, "status": "acknowledged"}
            
            if msg.reply:
                await self.nc.publish(msg.reply, json.dumps(response).encode())
                
        except Exception as e:
            print(f"Error handling command: {e}")
    
    async def _handle_event(self, msg):
        """Handle event messages"""
        try:
            data = msg.data.decode()
            print(f"Received event: {data}")
        except Exception as e:
            print(f"Error handling event: {e}")
    
    def publish_sync(self, subject: str, data: Dict[str, Any]) -> bool:
        """Publish message synchronously (thread-safe)"""
        if not self.nc or not self.loop:
            print("NATS not connected")
            return False
        
        try:
            # Schedule coroutine in the event loop
            future = asyncio.run_coroutine_threadsafe(
                self._publish_async(subject, data), 
                self.loop
            )
            future.result(timeout=5)  # Wait max 5 seconds
            return True
            
        except Exception as e:
            print(f"Error publishing to NATS: {e}")
            return False
    
    async def _publish_async(self, subject: str, data: Dict[str, Any]):
        """Async publish helper"""
        message = json.dumps(data).encode()
        await self.nc.publish(subject, message)
    
    def stop(self):
        """Stop NATS service"""
        print("Stopping NATS service...")
        self._running = False
        if self.executor:
            self.executor.shutdown(wait=True)
